### PR TITLE
Enhance plot_lowdim_rep to support string labels and robust NaN handling

### PR DESCRIPTION
**Pull request by CheeseLee888**
### Summary

This pull request enhanced the `plot_lowdim_rep` function to support:
String-based discrete labels (e.g., "A", "B", "C")
Robust handling of `NaN` values for continuous labels

And also added example notebook `test_plot_embeddings.ipynb` under `notebooks/`

### Notes
When plotting continuous labels with `NaN`, `NaN` points are visualized in green squares
For discrete labels, `NaN` is automatically treated as an additional category unless filtered

### TODO
Improve `NaN` handling for discrete labels: allow users to choose whether to plot or hide `NaN` points explicitly

### CODE REVIEW

File: contextualized/analysis/effects.py [hunk 1]
```diff
@@ -278,6 +279,16 @@ def plot_homogeneous_context_effects(
     Returns:
         None
     """
+
+    # if normalization is implemented, we hope to inverse_transform and show original context axis
+    if inverse_transform and getattr(model, "normalize", False):
+        if hasattr(model, "scaler_C") and model.scaler_C is not None:
+            # C = model.scaler_C.inverse_transform(C)
+            # C = inverse_transform_preserve_df(model.scaler_C, C)
```
Comment (cnellington) [inline contextualized/analysis/effects.py:287]: Various debugging comments like this still included. Only include explanatory comments, not comments intended for testing or switching behavior. Please remove and run black formatting. See CONTRIBUTING.md.

File: tests/test_normalize.py
Comment (cnellington) [inline tests/test_normalize.py:1]: We put unittests in the directories with the utils they are testing. Move to easy.

File: tests/test_normalize_wrapper.py
Comment (cnellington) [inline tests/test_normalize_wrapper.py:1]: Move to easy. This also seems like it supersedes the previous test. Please also translate any chinese comments to english.

File: tests/test_plot_homo_con_effects.py
Comment (cnellington) [inline tests/test_plot_homo_con_effects.py:1]: There are no assertions here. What is this testing?

**Review by cnellington — CHANGES_REQUESTED** [review]
Hi Peter, thanks for making this PR and being patient while I got to it. Left you some comments.

**Comment by CheeseLee888** [comment]
### Summary of changes

- Moved `test_normalize.py` and `test_normalize_wrapper.py` into `contextualized/easy/wrappers/` as they test wrapper-level logic.
- Clarified that `test_normalize_wrapper.py` was part of a failed attempt to support normalization in SKLearnWrapper. It is kept for reference.
- Added an explanatory comment and a minimal `assert` in `test_plot_homo_con_effects.py` to define it as a smoke test.
- Translated all Chinese comments to English.

Let me know if you'd prefer removing the legacy test file.

**Comment by cnellington** [comment]
@CheeseLee888 Please clarify which issue this PR is addressing, and remove anything outside of the issue scope. 

It sounds like there was an attempt to address #250, but if this cannot be addressed please add any info to the issue thread.

If this is addressing another issue, please clarify which issue and make sure the scope of the contributions is appropriate. Thanks for your help.

**Comment by cnellington** [comment]
Linking to #260 to track enhancement

File: contextualized/easy/ContextualizedRegressor.py
[the remark this answers is not in the document]
Comment (CheeseLee888) [reply contextualized/easy/ContextualizedRegressor.py:1]: I am still working on putting normalization into SkLearnWrapper, since it seems more complicated than having them in ContextualizedRegressor. I designed a test for normalization in SkLearnWrapper, but it turned out many errors. I had changed lots of codes beyond simply realizing normalization to fix those errors but still failed. To minimize disruption and keep the implementation self-contained, I opted to keep normalization logic within ContextualizedRegressor for now. However, we may also need normalization in Classifier or other models, then it might be better to put normalization in SkLearnWrapper.

File: tests/test_plot_homo_con_effects.py [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+import os
+import sys
+import numpy as np
+sys.path.insert(0, os.path.abspath(os.path.join(os.path.dirname(__file__), "..")))
+import pandas as pd
+from contextualized.easy import ContextualizedRegressor
+from contextualized.analysis.effects import plot_homogeneous_context_effects
+
+# Smoke test: check that plot_homogeneous_context_effects runs without error after denormalization
+def test_plot_homogeneous_context_effects():
```
[the remark this answers is not in the document]
Comment (CheeseLee888) [reply tests/test_plot_homo_con_effects.py:14]: Prof Lengerich once adviced me to add plot tests for normalized data(make normalized data go back to un-normalized then the plot axis is meaningful), so I included some codes about normalization in this file. Since now the two branch plot-embedding and normalization-clean is independent, I may not be able to run this test successfully. I may either wait for codes in normalization-clean to be merged or change this tests to a smaller scope without testing plotting normalized data.

File: tests/test_plot_homo_con_effects.py [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+    model.fit(C_df, X, Y)
+
+    # plot test for original C
+    print("🔍 Plotting with original (normalized internally) C...")
+    plot_homogeneous_context_effects(model, C_df, inverse_transform=True)
+
+    assert True  # ensures pytest treats this as a passing test
+
+if __name__ == "__main__":
+    test_plot_homogeneous_context_effects()
```
[the remark this answers is not in the document]
Comment (CheeseLee888) [reply tests/test_plot_homo_con_effects.py]: Romoved

File: tests/test_plot_homo_con_effects.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (CheeseLee888) [reply tests/test_plot_homo_con_effects.py]: Removed

File: pylint.svg
[the remark this answers is not in the document]
Comment (CheeseLee888) [reply pylint.svg:1]: Ran format sh but the score remained zero.

**Comment by CheeseLee888** [comment]
[discrete_labels_test.pdf](https://github.com/user-attachments/files/20743786/discrete_labels_test.pdf)
[continuous_labels_test_with_nan_plot_nan.pdf](https://github.com/user-attachments/files/20743787/continuous_labels_test_with_nan_plot_nan.pdf)
[continuous_labels_test_with_nan_hide_nan.pdf](https://github.com/user-attachments/files/20743788/continuous_labels_test_with_nan_hide_nan.pdf)
Plot test: string label and NaN control

**Comment by CheeseLee888** [comment]
Plot test: string label and NaN control
discrete_labels_test_page-0001
continuous_labels_test_with_nan_plot_nan_page-0001
continuous_labels_test_with_nan_hide_nan_page-0001

**Comment by CheeseLee888** [comment]
Put this file in Contextualized/
```python
import sys
import os
sys.path.append(os.path.abspath(".."))

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from Contextualized.contextualized.analysis.embeddings import plot_lowdim_rep


# set example
sample_N = 100
np.random.seed(0)
X = np.random.randn(sample_N, 2)

# -----------------------------
# 1. Discrete labels (categorical features)
# -----------------------------
labels_discrete = np.random.choice(["A", "B", "C"], size=sample_N) # Test with string labels

plot_lowdim_rep(
    low_dim=X,
    labels=labels_discrete,
    title="Discrete Labels Test",
    # cbar_label="",
    figname="discrete_labels_test",
)

# -----------------------------
# 2. Continuous labels (numerical features)
# -----------------------------
labels_cont = np.random.normal(loc=5.0, scale=2.0, size=sample_N)

plot_lowdim_rep(
    low_dim=X,
    labels=labels_cont,
    title="Continuous Labels Test",
    # cbar_label="",    
    figname="continuous_labels_test",
)

# -----------------------------
# 3. Too Many Discrete labels(automatically convert to continuous labels)
# -----------------------------
# labels_many = np.random.choice([f"Class_{i}" for i in range(15)], size=300)
labels_many = pd.Series(
    np.random.choice([f"Class_{i}" for i in range(15)], size=sample_N) # >10 categories => continuous
).astype("category").cat.codes

plot_lowdim_rep(
    low_dim=X,
    labels=labels_many,
    title="Too Many Discrete Labels Test",
    # cbar_label="",
    figname="Too_many_discrete_labels_test",
)

# -----------------------------
# 4. NaN values in discrete labels; simple, NaN regarded as another category
# -----------------------------
# labels_nan = labels_discrete.astype(object)
# labels_nan = pd.Series(np.random.choice(["A", "B", "C"], size=300)).astype("category").cat.codes.astype(object)
labels_nan = np.random.choice(["A", "B", "C"], size=sample_N)
labels_nan[::20] = np.nan  # put NaN in every 20th element

plot_lowdim_rep(
    low_dim=X,
    labels=labels_nan,
    title="Discrete Labels Test with NaN",
    # cbar_label="",
    figname="discrete_labels_test_with_nan",
)

# -----------------------------
# 5. NaN values in continuous labels;
# -----------------------------
labels_nan = pd.Series(np.random.normal(loc=5.0, scale=2.0, size=sample_N))
labels_nan[::20] = np.nan  # put NaN in every 20th element

plot_lowdim_rep(
    low_dim=X,
    labels=labels_nan,
    title="Continuous Labels Test with NaN, Plot NaN",
    # cbar_label="",
    figname="continuous_labels_test_with_nan_plot_nan",
    plot_nan=True
)

plot_lowdim_rep(
    low_dim=X,
    labels=labels_nan,
    title="Continuous Labels Test with NaN, Hide NaN",
    # cbar_label="",
    figname="continuous_labels_test_with_nan_hide_nan",
    plot_nan=False
)
```python

**Comment by CheeseLee888** [comment]
Pylint svg is fixed. Previously I ran it in Python 3.12 and something went wrong. I switched to Python 3.11 and turned out well.
